skorch/net.py

```diff
@@ -473,6 +476,45 @@ def initialize_module(self):
         self.module_ = module.to(self.device)
         return self
 
+    def _is_virtual_param(self, key):
+        return any(fnmatch.fnmatch(key, pat) for pat in self.virtual_params_)
+
+    def _virtual_setattr(self, param, val):
+        setattr(self, param, val)
+
+    def _register_virtual_param(self, param_patterns, fn=_virtual_setattr):
+        if not isinstance(param_patterns, list):
+            param_patterns = [param_patterns]
+        for pattern in param_patterns:
+            self.virtual_params_[pattern] = fn
+
+    def _apply_virtual_params(self, virtual_kwargs):
+        for pattern, fn in self.virtual_params_.items():
+            for key, val in virtual_kwargs.items():
+                if not fnmatch.fnmatch(key, pattern):
+                    continue
+                fn(key, val)
+
+    def initialize_params(self):
+        self.virtual_params_ = {}
+
+    def _set_lr_param(self, param, lr, optimizer_attr='optimizer_', optimizer_name='optimizer'):
```

Comment:
Since the virtual parameters may expand to objects, maybe `_set_lr_param` could be moved to `skorch.utils`. Otherwise, additional functions for other objects would expand the amount of code in `NeutralNet`. This would change the name to something like:

```python
def _set_optimizer_lr_param(net, param, lr, ...)
```

Do you see `optimizer_attr` and `optimizer_name` to be other values other than the default?

Reply:
> Since the virtual parameters may expand to objects, maybe `_set_lr_param` could be moved to `skorch.utils`. Otherwise, additional functions for other objects would expand the amount of code in `NeutralNet`. This would change the name to something like:
> 
> ```python
> def _set_optimizer_lr_param(net, param, lr, ...)
> ```

I think it might make sense to move these definitions to their own module since there may be more to come (criteria might want this ability, optimizer alone has several attributes like learning rate, momentum, averaging parameters).

> Do you see `optimizer_attr` and `optimizer_name` to be other values other than the default?

Yes, in the case when the user defines their own optimizer in `initialize_optimizer` this might happen.